torch/csrc/jit/codegen/cuda/lower_validation.cpp

```diff
@@ -518,85 +521,214 @@ void validateParallelize(Fusion* fusion) {
-          auto consumer_axis = *it;
+
+        producer_parallel_bitmap.set(producer_ptype);
+        producer_parallel_ids[getParallelTypeBitMapOffset(producer_ptype)] =
+            producer_axis;
+      }
+
+      for (auto consumer :
+           ir_utils::filterByType<TensorView>(expr->outputs())) {
+        // Stash information about parallelized consumer iteration domains
+        std::vector<IterDomain*> consumer_parallel_ids(
+            ParallelTypeBitmap::kNumParallelTypes, nullptr);
+        ParallelTypeBitmap consumer_parallel_bitmap;
+
+        for (const auto consumer_i : c10::irange(consumer->nDims())) {
+          auto consumer_axis = consumer->axis(consumer_i);
           auto consumer_ptype =
               par_map.getConcreteMappedID(consumer_axis)->getParallelType();
+
+          if (!isParallelTypeThread(consumer_ptype)) {
+            continue;
+          }
+
+          // When the consumer axis is a broadcast, it is not really
+          // parallelized unless thread-predicated and eventually concretized
+          if (consumer_axis->isBroadcast() &&
+              (!parallel_bcast_doms.get(consumer_ptype) ||
+               !GpuLower::current()
+                    ->concretizedBroadcastDomains()
+                    .isConcretized(consumer_axis))) {
+            continue;
+          }
+
+          consumer_parallel_bitmap.set(consumer_ptype);
+          consumer_parallel_ids[getParallelTypeBitMapOffset(consumer_ptype)] =
+              consumer_axis;
+        }
+
+        for (auto parallel_type : kParallelTypeThreads) {
+          bool valid_parallel_comm = true;
+
+          auto parallel_type_i = getParallelTypeBitMapOffset(parallel_type);
+
+          auto p_id = producer_parallel_ids[parallel_type_i];
+          auto c_id = consumer_parallel_ids[parallel_type_i];
+
+          if (p_id == nullptr && c_id == nullptr) {
+            continue;
+          } else if (p_id != nullptr && c_id != nullptr) {
+            // If same parallel type and mapped, no need for syncs unless
+            // producer is in smem, producer parallel type is a thread
+            // dimension, and consumer concretizes the dimension. This sync is
+            // due to the redundant predicate omission in lower thread
+            // predicate.
+            if (loop_map.areMapped(p_id, c_id)) {
+              if (p_id->isBroadcast() && !c_id->isBroadcast() &&
+                  isParallelTypeThreadDim(parallel_type)) {
+                raw_dims.set(parallel_type);
+              }
+              auto halo_info = GpuLower::current()->haloInfo();
+              if (halo_info.hasHaloWidth(p_id) !=
+                  halo_info.hasHaloWidth(c_id)) {
+                raw_dims.set(parallel_type);
+              }
+              if (halo_info.hasHaloWidth(p_id)) {
+                if (!GpuLower::current()->haloInfo().extentEqual(p_id, c_id)) {
+                  raw_dims.set(parallel_type);
+                }
+              }
```

Comment:
Don't we need to validate the producer memory type is smem?

Reply:
I guess we don't need to validate it's smem, but we just need to do a RAW sync when the producer is on smem, but that's being taken care by the sync inserter, so adding to the `raw_dims` set no matter it's smem works fine.

Reply:
For the halo part, it could be also done as:

```
if (halo_info.hasHaloWidth(p_id) != halo_info.hasHaloWidth(c_id) ||
   (halo_info.hasHaloWidth(p_id) && halo_info.hasHaloWidth(c_id) && halo_info.getHaloWidth(p_id) != halo_info.getHaloWidth(c_id))) {
  raw_dims.set(parallel_type);
}
```

Reply:
I think you're right and believe this is now fixed.